Use static member functions from class instead of instances

aureleoules opened the pull request:
I noticed while reviewing a PR that some static class member functions are accessed by instance instead of class. I believe this makes the code misleading.
This PR enables the clang-tidy check 'readability-static-accessed-through-instance'. 
> Checks for member expressions that access static members through instances, and replaces them with uses of the appropriate qualified-id.

https://clang.llvm.org/extra/clang-tidy/checks/readability/static-accessed-through-instance.html


I used the `-fix` option of clang-tidy to discover and fix these issues.

w0xlt reviewed (commented):
Approach ACK

MarcoFalke commented on src/crypto/muhash.cpp:
```diff
@@ -299,7 +299,7 @@ Num3072 MuHash3072::ToNum3072(Span<const unsigned char> in) {
 
     uint256 hashed_in{(HashWriter{} << in).GetSHA256()};
-    ChaCha20(hashed_in.data(), hashed_in.size()).Keystream(tmp, Num3072::BYTE_SIZE);
+    ChaCha20(hashed_in.data(), uint256::size()).Keystream(tmp, Num3072::BYTE_SIZE);
```
I think the pattern is generally `a.data(), a.size()`. If you want to change this, it might be best to use spans instead of the changes here in this pull.

aureleoules replied:
i am unfamiliar with spans, how would that work?

MarcoFalke replied:
See `std::span` and our implementation of it `Span`

aureleoules replied:
Thanks, attempted in b8f72010a424c48cabc92d8f0e7a43e8eb8460e8.
Not sure if I should add this commit in a separate PR.

aureleoules committed:
crypto: Add Write function overload with Span

This simplifies the usage of various hasher Write functions as there is
no need to pass the buffer size.

MarcoFalke commented on src/core_write.cpp:
```diff
@@ -69,7 +69,7 @@ std::string FormatScript(const CScript& script)
         break;
     }
-    return ret.substr(0, ret.empty() ? ret.npos : ret.size() - 1);
+    return ret.substr(0, ret.empty() ? std::string::npos : ret.size() - 1);
```
personally I prefer ret.npos, or at least think that either is fine and we don't need a linter for this

aureleoules replied:
dropped e16f6a34b5a200e3ab163dd935f26f17bd1c172f and rolled back to `ret.npos` instead of `std::string::npos` in ff9c66a11d5d2c4dc64d026a0719cb3fc586d17d

MarcoFalke replied:
I only left a comment on this line, but I meant all other lines as well.

I really don't get the recent rush to enforce meaningless style decisions with clang-tidy, even happily introducing bugs along the way (https://github.com/bitcoin/bitcoin/pull/25695#discussion_r954086171).

aureleoules replied:
Yes I meant I updated it everywhere as well.

Well, it came from a readability issue and so I thought it would be better to enforce it.

aureleoules replied:
Happy to drop ff9c66a11d5d2c4dc64d026a0719cb3fc586d17d and keep 6f8c291218041ae843dd9099c7103e607d0f8a39 if needed

aureleoules committed:
refactor: Use static member functions instead of instance

ajtowns commented:
+1 on switching to Span, -1 on switching from the instance to the class for accessing static methods though.

luke-jr commented:
>I noticed while reviewing a PR that some static class member functions are accessed by instance instead of class. I believe this makes the code misleading.

I don't agree. Sometimes it makes sense to access by instance instead of class.

If there's particular cases that are confusing, those can be fixed, but I don't think adopting this as a hard rule is a good idea.
